File: sdks/js/packages/spark-sdk/src/utils/proof.ts
import { sha256 } from "@scure/btc-signer/utils";

export function proofOfPossessionMessageHashForDepositAddress(
  userPubkey: Uint8Array,
  operatorPubkey: Uint8Array,
  depositAddress: string,
): Uint8Array {
  const encoder = new TextEncoder();
  const depositAddressBytes = encoder.encode(depositAddress);

  const proofMsg = new Uint8Array([
    ...userPubkey,
    ...operatorPubkey,
    ...depositAddressBytes,
  ]);
  return sha256(proofMsg);
}

File: spark/common/proof.go
package common

import (
	"bytes"
	"crypto/sha256"
)

// ProofOfPossessionMessageHashForDepositAddress generates a hash of the proof of possession message for a deposit address.
func ProofOfPossessionMessageHashForDepositAddress(userPubkey, operatorPubkey, depositAddress []byte) []byte {
	proofMsg := bytes.Join([][]byte{
		userPubkey,
		operatorPubkey,
		depositAddress,
	}, nil)
	hash := sha256.Sum256(proofMsg)
	return hash[:]
}
